agent/engine/docker_task_engine.go

```diff
@@ -248,9 +251,72 @@ func (engine *DockerTaskEngine) Init(ctx context.Context) error {
 	// Now catch up and start processing new events per normal
 	go engine.handleDockerEvents(derivedCtx)
 	engine.initialized = true
+	go engine.startPeriodicExecAgentsMonitoring(derivedCtx, monitorExecAgentsInterval)
 	return nil
 }
 
+func (engine *DockerTaskEngine) startPeriodicExecAgentsMonitoring(ctx context.Context, monitorInterval time.Duration) {
+	engine.monitorExecAgentsTicker = time.NewTicker(monitorInterval)
+	for {
+		select {
+		case <-engine.monitorExecAgentsTicker.C:
+			go engine.monitorExecAgentProcesses(ctx)
+		case <-ctx.Done():
+			engine.monitorExecAgentsTicker.Stop()
+			return
+		}
+	}
+}
+
+func (engine *DockerTaskEngine) monitorExecAgentProcesses(ctx context.Context) {
+	// TODO: [ecs-exec]add jitter between containers to not overload docker with top calls
+	engine.tasksLock.RLock()
+	defer engine.tasksLock.RUnlock()
+	for _, mTask := range engine.managedTasks {
+		task := mTask.Task
+		if !task.IsExecCommandAgentEnabled() {
+			continue
+		}
+		for _, c := range task.Containers {
+			go engine.monitorExecAgentRunning(ctx, task.Arn, c)
+		}
+	}
+}
+
+func (engine *DockerTaskEngine) monitorExecAgentRunning(ctx context.Context,
+	tArn string, c *apicontainer.Container) {
+	if c.GetExecCommandAgentMetadata() == nil {
+		return
+	}
+	pID := c.GetExecCommandAgentMetadata().PID
+	cName := c.Name
+	cID := c.RuntimeID
+	if cID == "" || pID == "" {
+		return
+	}
+	resp, err := engine.client.TopContainer(ctx, cID, dockerclient.TopContainerTimeout, pID)
+	if err != nil {
+		if _, ok := err.(*dockerapi.DockerTimeoutError); ok {
+			seelog.Warnf("Task engine [%s]: Timed out monitoring ExecCommandAgent process for container %s", tArn, cName)
+			return
+		}
+		if strings.Contains(err.Error(), dockerapi.TopProcessNotFoundErrorName) {
+			seelog.Errorf("Task engine [%s]: ExecCommandAgent Process for container %s not running, restarting", tArn, cName)
+			c.SetExecCommandAgentMetadata(&apicontainer.ExecCommandAgentMetadata{})
+			// TODO: [ecs-exec]restart exec agent process for this container
+			return
+		}
+		seelog.Errorf("Task engine [%s]: Error finding ExecCommandAgent process for container %s: %v", tArn, cName, err)
```

Comment:
Maybe we should consider retrying here, just in case the error is transient? Just something to consider, but not a blocker for this PR.

Reply:
was thinking it will be covered in the next monitor cycle? but yes, I shall add a TODO to consider this.

Reply:
Yeah -- I see this check process as two-fold: 
if we don't see the execute command agent running based on docker top and the `ExecCommandAgentMetadata.PID`, then I'd like to ask docker exec inspect for the exit code of that process and act accordingly.
I think we should only be restarting for a subset of exit codes (maybe just 2?), else we should log an error...